### PR TITLE
Add delete usecase and GraphQL API for ChangeLogs

### PR DESCRIPTION
**Pull request by rohithbalaji123**
Part of #612 

### Description
Add delete usecase along with GraphQL API.

The response given by the graphQL API is,
```
{
  "data": {
    "authMutation": {
      "deleteChange": "<delete-id>"
    }
  }
}
```

### CODE REVIEW

File: backend/app/adapter/gqlapi/resolver/authmutation.go [hunk 1]
```diff
@@ -83,6 +88,16 @@ func (a AuthMutation) CreateChange(args *CreateChangeArgs) (Change, error) {
...
 }
 
+// DeleteChange removes a Change in the change log
+func (a AuthMutation) DeleteChange(args *DeleteChangeArgs) (*string, error) {
+	ID := args.ID
+	err := a.changeLog.DeleteChange(ID)
+	if err != nil {
+		return nil, err
+	}
+	return &ID, nil
```
Comment (arberiii) [inline backend/app/adapter/gqlapi/resolver/authmutation.go:98]:
```suggestion
	err := a.changeLog.DeleteChange(args.ID)
	if err != nil {
		return nil, err
	}
	return &args.ID, nil
```
I don't see why we can't pass the args.ID directly

File: backend/app/usecase/changelog/changelog.go [hunk 2]
```diff
@@ -17,6 +17,7 @@ type ChangeLog interface {
 	GetChangeLog() ([]entity.Change, error)
 	GetLastViewedAt(user entity.User) (*time.Time, error)
 	ViewChangeLog(user entity.User) (time.Time, error)
+	DeleteChange(ID string) error
```
Comment (arberiii) [inline backend/app/usecase/changelog/changelog.go:20]:
```suggestion
	DeleteChange(id string) error
```

File: backend/app/usecase/changelog/changelog.go [hunk 3]
```diff
@@ -84,6 +85,11 @@ func (p Persist) ViewChangeLog(user entity.User) (time.Time, error) {
 	return time.Time{}, err
 }
 
+// DeleteChange removes the change having given ID
+func (p Persist) DeleteChange(ID string) error {
```
Comment (arberiii) [inline backend/app/usecase/changelog/changelog.go:89]:
```suggestion
func (p Persist) DeleteChange(id string) error {
```

File: backend/app/adapter/gqlapi/resolver/authmutation.go [hunk 4]
```diff
@@ -83,6 +88,16 @@ func (a AuthMutation) CreateChange(args *CreateChangeArgs) (Change, error) {
 	return newChange(change), err
 }
 
+// DeleteChange removes a Change in the change log
```
Comment (arberiii) [inline backend/app/adapter/gqlapi/resolver/authmutation.go:91]: Should we emphasize that the changelog with the given ID is deleted?

File: backend/app/usecase/changelog/changelog_test.go [hunk 5]
```diff
@@ -344,3 +344,80 @@ func TestPersist_ViewChangeLog(t *testing.T) {
...
+	testCases := []struct {
+		name                  string
+		changeLog             []entity.Change
+		deleteChangeId        string
+		expectedChange        []entity.Change
+		expectedChangeLogSize int
+		hasErr                bool
+	}{
+		{
+			name: "delete change successfully",
```
Comment (arberiii) [inline backend/app/usecase/changelog/changelog_test.go:362]: Can you add a test case where the deleteChangeId doesn't exist?

File: backend/app/usecase/changelog/changelog_test.go [hunk 6]
```diff
@@ -344,3 +344,80 @@ func TestPersist_ViewChangeLog(t *testing.T) {
...
+			)
+
+			err = persist.DeleteChange(testCase.deleteChangeId)
+			if testCase.hasErr {
+				assert.NotEqual(t, nil, err)
+				return
+			}
+			assert.Equal(t, nil, err)
+
+
```
Comment (arberiii) [inline backend/app/usecase/changelog/changelog_test.go:414]:
```suggestion
```

File: backend/app/adapter/gqlapi/resolver/authmutation.go (hunk 4 above)
Comment (byliuyang) [inline backend/app/adapter/gqlapi/resolver/authmutation.go:91]:
```suggestion
// DeleteChange removes a Change with given ID from change log
```

File: backend/app/usecase/changelog/changelog.go [hunk 7]
```diff
@@ -84,6 +85,11 @@ func (p Persist) ViewChangeLog(user entity.User) (time.Time, error) {
...
+	return p.changeLogRepo.DeleteChange(ID)
```
Comment (byliuyang) [inline backend/app/usecase/changelog/changelog.go:90]:
```suggestion
	return p.changeLogRepo.DeleteChange(id)
```

File: backend/app/usecase/changelog/changelog_test.go [hunk 8]
```diff
@@ -344,3 +344,80 @@ func TestPersist_ViewChangeLog(t *testing.T) {
...
+func TestPersist_DeleteChange(t *testing.T) {
+	t.Parallel()
+
+	summaryMarkdown1 := "summary 1"
+	summaryMarkdown2 := "summary 2"
+	testCases := []struct {
+		name                  string
+		changeLog             []entity.Change
+		deleteChangeId        string
+		expectedChange        []entity.Change
```
Comment (byliuyang) [inline backend/app/usecase/changelog/changelog_test.go:357]:
```suggestion
		expectedChangeLog        []entity.Change
```

File: backend/app/usecase/changelog/changelog_test.go [hunk 9]
```diff
@@ -344,3 +344,80 @@ func TestPersist_ViewChangeLog(t *testing.T) {
...
+			}
+			assert.Equal(t, nil, err)
+
+
+			changeLog, err := persist.GetChangeLog()
+			assert.Equal(t, testCase.expectedChange, changeLog)
+
+			assert.Equal(t, nil, err)
+
+			assert.Equal(t, testCase.expectedChangeLogSize, len(changeLog))
```
Comment (byliuyang) [inline backend/app/usecase/changelog/changelog_test.go:420]:
```suggestion
                         assert.Equal(t, nil, err)
			assert.SameElements(t, testCase.expectedChangeLog, changeLog)
```
If the change logs are the same, there is no point of assessing length.

File: backend/app/adapter/gqlapi/schema.go [hunk 10]
```diff
@@ -35,6 +35,7 @@ type Change {
 type AuthMutation {
 	createShortLink(shortLink: ShortLinkInput!, isPublic: Boolean!): ShortLink
 	createChange(change: ChangeInput!): Change!
+	deleteChange(ID: String!): String
```
Comment (byliuyang) [inline backend/app/adapter/gqlapi/schema.go:38]:
```suggestion
	deleteChange(id: String!): String
```

File: backend/app/usecase/changelog/changelog.go [hunk 11]
```diff
@@ -84,6 +85,11 @@ func (p Persist) ViewChangeLog(user entity.User) (time.Time, error) {
 	return time.Time{}, err
 }
 
+// DeleteChange removes the change having given ID
```
Comment (byliuyang) [inline backend/app/usecase/changelog/changelog.go:88]:
```suggestion
// DeleteChange removes the change with given ID
```

**Pull request merged**